ADAL/tests/unit/ADTokenCacheItemTests.m

```diff
@@ -43,110 +43,480 @@ - (void)tearDown
+    XCTAssertNotEqualObjects(lhs, rhs);
+}
+
+- (void)testIsEqual_whenUserInformationIsEqual_shouldReturnTrue
+{
+    ADTokenCacheItem *lhs = [ADTokenCacheItem new];
+    lhs.userInformation = [self adCreateUserInformation:TEST_USER_ID];
+    ADTokenCacheItem *rhs = [ADTokenCacheItem new];
+    rhs.userInformation = [self adCreateUserInformation:TEST_USER_ID];
+    
+    XCTAssertEqualObjects(lhs, rhs);
+}
+
+- (void)testIsEqual_whenSessionKeyIsNotEqual_shouldReturnFalse
+{
+    ADTokenCacheItem *lhs = [ADTokenCacheItem new];
+    lhs.sessionKey = [@"qwe" dataUsingEncoding:NSUTF8StringEncoding];
+    ADTokenCacheItem *rhs = [ADTokenCacheItem new];
+    rhs.sessionKey = [@"asd" dataUsingEncoding:NSUTF8StringEncoding];
+    
+    XCTAssertNotEqualObjects(lhs, rhs);
+}
+
+- (void)testIsEqual_whenSessionKeyIstEqual_shouldReturnTrue
+{
+    ADTokenCacheItem *lhs = [ADTokenCacheItem new];
+    lhs.sessionKey = [@"session key" dataUsingEncoding:NSUTF8StringEncoding];
+    ADTokenCacheItem *rhs = [ADTokenCacheItem new];
+    rhs.sessionKey = [@"session key" dataUsingEncoding:NSUTF8StringEncoding];
+    
+    XCTAssertEqualObjects(lhs, rhs);
+}
+
+- (void)testIsEqual_whenTombstoneIsNotEqual_shouldReturnFalse
+{
+    ADTokenCacheItem *lhs = [ADTokenCacheItem new];
+    [lhs setValue:@{@"some key": @"some value"} forKey:@"tombstone"];
+    ADTokenCacheItem *rhs = [ADTokenCacheItem new];
+    [rhs setValue:@{@"some key 2": @"some value 2"} forKey:@"tombstone"];
+    
+    XCTAssertNotEqualObjects(lhs, rhs);
+}
+
+- (void)testIsEqual_whenTombstoneIsEqual_shouldReturnTrue
+{
+    ADTokenCacheItem *lhs = [ADTokenCacheItem new];
+    [lhs setValue:@{@"some key": @"some value"} forKey:@"tombstone"];
+    ADTokenCacheItem *rhs = [ADTokenCacheItem new];
+    [rhs setValue:@{@"some key": @"some value"} forKey:@"tombstone"];
+    
+    XCTAssertEqualObjects(lhs, rhs);
+}
+
+- (void)testIsEqual_whenAdditionalClientIsNotEqual_shouldReturnFalse
+{
+    ADTokenCacheItem *lhs = [ADTokenCacheItem new];
+    lhs.additionalClient = [@{@"k1":@"v1"} mutableCopy];
+    ADTokenCacheItem *rhs = [ADTokenCacheItem new];
+    rhs.additionalClient = [@{@"k2":@"v2"} mutableCopy];
+    
+    XCTAssertNotEqualObjects(lhs, rhs);
+}
+
+- (void)testIsEqual_whenAdditionalClientIsEqual_shouldReturnTrue
+{
+    ADTokenCacheItem *lhs = [ADTokenCacheItem new];
+    lhs.additionalClient = [@{@"k1":@"v1"} mutableCopy];
+    ADTokenCacheItem *rhs = [ADTokenCacheItem new];
+    rhs.additionalClient = [@{@"k1":@"v1"} mutableCopy];
+    
+    XCTAssertEqualObjects(lhs, rhs);
+}
+
+- (void)testIsEqual_whenAdditionalServerIsNotEqual_shouldReturnFalse
+{
+    ADTokenCacheItem *lhs = [ADTokenCacheItem new];
+    [lhs setValue:@{@"k1":@"v1"} forKey:@"additionalServer"];
+    ADTokenCacheItem *rhs = [ADTokenCacheItem new];
+    [rhs setValue:@{@"k2":@"v2"} forKey:@"additionalServer"];
+    
+    XCTAssertNotEqualObjects(lhs, rhs);
+}
+
+- (void)testIsEqual_whenAdditionalServerIsEqual_shouldReturnTrue
```

Comment:
There are 2 types of tests:
- Verify that objects are equal when all properties are set:
`testIsEqual_whenAllPropertiesAreEqual_shouldReturnTrue`
- Verify that objects are equal when all properties are nil and only one property is set, for example:
```
testIsEqual_whenResourceIsEqual_shouldReturnTrue
testIsEqual_whenResourceIsNotEqual_shouldReturnFalse
```
We can add other tests that verify mix of nil and not nil properties, but I think those test will overlap with current ones.